src/wasm/wat-parser-internal.h

```diff
@@ -0,0 +1,521 @@
+/*
+ * Copyright 2022 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Usage note
+// ----------
+//
+// This parser is a work in progress and this file should not yet be included
+// anywhere except for in its own tests. Once the parser is usable, we will add
+// wat-parser.h to declare the public parsing API and wat-parser.cpp to
+// implement the public parsing functions in terms of the private API in this
+// header. The private API will stay in this header rather than moving to
+// wat-parser.cpp so that we can continue to unit test it.
```

Comment:
Ideally I would have this entire file be a .cpp file, since it is all private details of the future parser API and everything including the declarations is in an anonymous namespace. The only reason there is a header at all is so that I can include the private API into the unit test source.

An alternative would be to put the `Lexer` and `Token` definitions into a header since those are the things that are used in the tests, then everything else could go in a .cpp file.

Reply:
Yeah, I see why classes should be in a header file. I was asking about method implementation that are outside of class declarations. But we don't have many of them anyway because most method impls are class inline. I think this should be fine at this point.

Reply:
Right, logically I want to treat this entire file as a single .cpp file including all the declarations and definitions. The fact that it's actually a header is basically a hack for testing.